doc/decisions/library_directory_structure.md

```diff
@@ -2,21 +2,67 @@
 
 ## Problem
 
-Currently src/libs/elektra contains the source of several libraries. Each of the libraries should have its own folder.
+Currently `src/libs/elektra` contains the source of several libraries. Each of the libraries should have its own folder.
 Furthermore, we need to allow rust source to exist in parallel.
 
 ## Constraints
 
+- It must be possible to set up include-paths via CMake so that `#include "header.h"` works the same in the build directory and when installed.
+
 ## Assumptions
 
 ## Considered Alternatives
 
 ## Decision
 
+We use this directory tree (YAML syntax for highlighting):
+
+```yaml
+doc:
+examples:
+scripts:
+src:
+ c:
+  libs: one folders per library written in C, contains headers and source
+  plugins: one folder per plugin written in C
+  tools: one folder per tool written in C
+  tests: tests written in C
+ rust:
+  binding: Rust binding of C API
+  libs:
+   core: Rust alternative of libelektra-core
+   _: {other libraries written in Rust}
+  _: {other folders as needed for Rust, unit tests are normally in the same file as source}
+ cpp:
+  binding: C++ binding of C API
+  libs: one folder per lib, contains headers and source, written in C++
+  plugins: one folder per plugin, only plugins written in C++
+  tools: one folder per tool, only tools written in C++
+  tests: tests written in C++
+ _: {
+         folders for other languages,
+         "binding" is always the folder for the binding to the C API,
+         rest depends on the language
+    }
+```
```

Comment:
@lawli3t @tucek Some suggestions for the Rust and Java parts of the directory tree would be appreciated.

Reply:
It is a clean separation but I think it is not so helpful for beginners who have no idea in which language we implemented which part. So what about the other way round (put the programming language in the deepest level of nesting)?

```yaml
src:
  libs: one folder per lib
    core:  for libelektra-core, other libs are separated
       include: only the C headers, as the other headers would be in "bindings"
       src:
          c: src in C
          rust: source in Rust
       tests: tests written in any language
          c: tests in C
          rust: tests in Rust
```

if there is only one implementation language, no c/rust/... subfolders are needed.

Reply:
I don't like the separate `include` folder, it just complicates things. Having the header next to the code file makes browsing the repo much easier and `#include "foo.h"` in `foo.c` is simpler than `#include <foo.h>` and hope include paths are set correctly.

Putting the language at the bottom also makes it much harder to have per-language CMake code. In my example, `src/rust/CMakeLists.txt` would do all the checks for Rust (i.e. is `cargo` installed, do we have the required Rust version, etc.). Only if those checks succeed, will the subfolders be included.
With your system you have to (at the very least) call some kind of check function in every `.../rust/CMakeLists.txt`.

Probably more importantly most languages won't actually use CMake directly as the build tool. Java uses Gradle, Rust uses Cargo, etc. If there isn't a single root directory for all Java/Rust code, you need to setup multiple separate projects for these build tools instead of just a single one with sub-projects/modules.

> it is not so helpful for beginners who have no idea in which language we implemented which part

This could simply be solved by adding a `src/README.md` file similar to what we have now in `src/plugins/README.md`.

I also think my structure has a different advantage for beginners. Your version is better for "I want to know about the code of `libelektra-core.so`". Mine is easier for "I know Rust, and what to contribute to Elektra". In your version a beginner would have to check every library, plugin, tool, etc. for a `rust` folder. In my structure, they just open `src/rust` and see what is there.

If you really want the language at the bottom, I would suggest:

```yaml
src:
 libs:
  core-c: C implementation of libelektra-core.so
  core-rust: Rust implementation of libelektra-core.so
  opts: libelektra-opts.so has only a single implementation, so no suffix
  # ...
 # plugins and tools use the same suffix idea as above
 plugins:
 tools:
 bindings:
  rust: Rust bindings
  cpp: C++ bindings
  java: Java bindings
```

However, I still don't like this structure because of the build system (CMake, Gradle, Cargo, ...) disadvantages.

Reply:
I would also very much prefer having everything related to one language in one folder, it would probably mess with cargo as well if everything was scattered among multiple different folders. Going by the recommended [Rust project structure](https://doc.rust-lang.org/cargo/guide/project-layout.html), most of the stuff should be contained in one parent folder as cargo kind of expects you to adhere to the structure (it doesn't necessitate it but it certainly makes things a lot easier, especially in CI)

I think we should be just fine with the binding and libs folder, tests should be contained in the respective folders anyway. The projects contained in those folder should try to adhere to the structure mentioned above.

It also makes sense from a developers point of view because mostly I will be working on one single project and then have all the stuff related to that in one folder (which i can declare as root folder in the IDE). Other setups would be cumbersome to manage and would require specific configuration for every Editor/IDE....

Reply:
Gradle supports, creating subprojects with an arbitrary directory layout (see [this SO post](https://stackoverflow.com/questions/20807490/defining-subprojects-in-a-deeply-nested-directory-tree)). Adding dependencies to individual gradle project in CMake might be complicated, but I'm not sure this will ever be needed. Our Java code will probably not be called from a non-JVM language. Therefore, we can simply use Gradle to build everything that depends on the Java code.

@lawli3t For Rust, [this](https://github.com/corrosion-rs/corrosion) looks very helpful.

In general for a multi-language project like Elektra, [Bazel](https://bazel.build/) probably would be a better choice than CMake. But I have no experience with Bazel and I certainly don't think anybody is going to rewrite all our CMake code for Bazel.

Reply:
See https://github.com/ElektraInitiative/libelektra/pull/4243/files#r807268499 for further discussion.

Reply:
I like the idea of having this mixed layout but using the whole src folder for Java (Gradle) and Rust (Cargo) while having tools/plugins/libs in their proper place. Still open is if PLUGINS can then actually be used for Java/Rust plugins.

Btw. in theory CMake (like Bazel) can also build other languages (we actually used the add_jar already), practically Java programmers want Gradle and Rust programmers want Cargo. Bazel won't help.

Reply:
> Still open is if PLUGINS can then actually be used for Java/Rust plugins.

Do you mean the CMake variable `PLUGINS`?

Reply:
Exactly.

Reply:
I don't see why passing PLUGINS to Gradle/Cargo could be a big problem. E.g. you can easily extend Cargo.toml from within CMake and pass anything as you want.

> How to pick between alternative implementations?

I am afraid we need a new CMake variable for that.

> How to express dependencies? e.g. a library needs core but doesn't care which implementation

Then you obviously do not need to specify anything as it should be a full replacement.

> or a Rust library needs a library that only exists in C (so cargo doesn't know how to build it)

I don't think this discussion is helpful for @lawli3t, let us discuss problems he actually has.

Reply:
> > How to express dependencies? e.g. a library needs core but doesn't care which implementation
> 
> Then you obviously do not need to specify anything as it should be a full replacement.

Lets's say `libelektra-foo` depends on `core`, but doesn't care about the implementation. What CMake dependencies should `libelektra-foo` now define?

I think we need a dummy target `libelektra-core` that depending on other configuration either depends on `libelektra-core-c` or on `libelektra-core-rust`.

> you can easily extend Cargo.toml from within CMake and pass anything as you want.

I'm not sure what you mean. What we should absolutely avoid (and also remove where it exists already) is something like having a `Cargo.toml.in` that needs to be processed by CMake's `configure_file` first. This makes it very annoying or even impossible to use IDEs that don't know about CMake.

> I don't see why passing PLUGINS to Gradle/Cargo could be a big problem.

The problem is who determines, if all plugin dependencies of a plugin are present. If plugin `foo` needs plugin `bar`, but one is written in C and the other in Java this is not so easy. It's doable, but probably very annoying to implement. CMake needs to ask Gradle/Cargo for all the plugins (and their dependencies) that can be built with these tools. Alternatively, we need to reimplement the dependency checks in Gradle/Cargo and pass the list of enabled plugins. But this still need two-way communication, because Gradle/Cargo could decide to disable a plugin because of a missing dependency, which in turn could lead to CMake having to disable another plugin, etc.

Whatever we decided to do, we need some interoperation between CMake and Gradle/Cargo.

> I don't think this discussion is helpful for lawli3t, let us discuss problems he actually has.

Neither this PR nor this thread are about specific problems somebody had. This is about a general decision. A decision that has very far reaching consequences, so we should carefully consider all possible problems. Once we decide on a directory structure, I don't think anybody will want to invest the time change it again in the near future.

Reply:
> Once we decide on a directory structure, I don't think anybody will want to invest the time change it again in the near future.

I totally agree but I am afraid that we can discuss a lot and it will never work out that way unless shown with a "prototype". Let us see how #4341 works out (and help in the problems coming up there), and then let us continue finding a solution (if there are still structural problems).

Reply:
AFAICT the most current discussion on the actual library structure is in https://github.com/ElektraInitiative/libelektra/pull/4243/files#r851730391. So if you want to add something, please continue the discussion there.

Regarding "it will never work out that way": The misunderstanding is probably because it has been months since the original discussion, but my original comment was exactly in the spirit of "it never works out that way". I was anticipating a potential problem, and asked about it (https://github.com/ElektraInitiative/libelektra/pull/4223#discussion_r809951795, "How to express dependencies?"). Your response was "That's not helpful right now, please discuss the problems that currently exist" (paraphrased). To which I said, "Potential future problems that come from this change are relevant right now, because changing this again is a lot of work" (paraphrased). So I'm not really sure, what you wanted to say here.

Anyway, since #4341 doesn't actually need to change the structure of the whole repo, I'm fine with postponing this discussion until it is merged.